test/test_xla_sharding.py

```diff
@@ -175,6 +175,17 @@ def test_model_weight_metrics(self):
     self.assertIn("VirtualDeviceUsage", met.counter_names())
     self.assertNotEqual(met.counter_value("VirtualDeviceUsage"), 0)
 
+  def test_no_sharding(self):
```

Comment:
I need to do some catch up, so when `XLA_USE_SPMD` is set, will `xla_device` return a virtual device? Is this test trying to test operations on real device or virtual device without sharding?

Reply:
Yeah that flag enables the virtual device.

The goal of this test is to define tensors on the virtual device, and then do an operation without applying sharding, i.e. without transferring data to the TPU. We want to make sure that tensors on the virtual device can be used without being sharded, and that there aren't any data transfer issues.

Reply:
make sense, would the computation
```
t3 = t1 + t2
```
being run on all devices but replicated?

Reply:
Confirmed that the computation happens on device `TPU:0` observing logs in `third_party/xla_client/pjrt_computation_client.cc`, but there is no replication or sharding going on.

```
start no sharding test
2022-12-18 22:41:59.286360: I   81440 tensorflow/compiler/xla/xla_client/pjrt_computation_client.cc:279] Executing PjRt computation on TPU:0
2022-12-18 22:41:59.286452: I   81440 tensorflow/compiler/xla/xla_client/pjrt_computation_client.cc:283] Executing PjRt computation on TPU:0
2022-12-18 22:41:59.287176: I   81440 tensorflow/compiler/xla/xla_client/pjrt_computation_client.cc:326] Returning 1 results
end no sharding test
```

Reply:
Virtual device abstraction won't trigger SPMD (sharded computation). This is correct.